docs/DeveloperGuide.md

```diff
@@ -1,13 +1,304 @@
+  * Disadvantage: If the program crashes or exits incorrectly, data will not be saved.
+* Saving the data after every command:
+  * Advantage: Changes are saved after every command.
+  * Disadvantage: Executing command might slow down the program when there is a large amount of data to be saved.
+
+### Visualization Feature 
+
+This feature is implemented with the help of [XChart](https://knowm.org/open-source/xchart/), a simple charting library for Java by Knowm.
+
+By typing in the vis command with the appropriate arguments (`/s` and `/t`), users will be able to visualize their 
+income or expense using visualization tools (Piechart, Bar Chart or Radar Chart)
+
+Demo: 
+
+`vis /t expense /c pie`
+
+Output
+
+`Displaying piechart for expense`
+A message will be shown telling you that the chart is being displayed
+
+![](images/vis/visOutput.png)
+
+This feature was implemented with the help of three different classes.
+They are namely: Visualizer, Categorizer, VisCommand (Inherits from abstract Command Class)
+
+VisCommand's Role: 
+1) Read the parameters of the vis command entered by the user
+- `/t` Reads the type of cashflow that the user wants to visualize (income/expense)
+- `/c` Reads the type of visualization tools the user wants (piechart/barchart)
+
+2) Calls the Cateorgizer to sort cashflow (Income/Expense) according to type
+
+3) Calls the Visualizer to display the chart to the user
+
+Categorizer's Role: 
+
+According to the cashflow type (Income/Expense) arugment passed in, the Categorizer sorts the 
+specified cashflow entry according to type using a Hashmap which is returned and used by the Visualizer
+
+Visualizer's Role: 
+
+According to the chart type (Pie/Bar/Radar) argument and the Hashmap obtained from the categorizer passed in, 
+the visualizer displays the specified visualization chart by calling the charting library Xchart.
+
+### Class Diagram
+
+![](images/vis/visualisationClass.png)
+
+### Sequence Diagram 
+
+Overall 
+
+![](images/vis/visualisationSequence.png)
+
+Categorizer
+
+![](images/vis/categorizerSequence.png)
+
+Visualizer
+
+![](images/vis/visualizerSequence.png)
+
+### Add Income/Expense Feature
+
+The add income/expense command has 2 compulsory arguments `/t` and `/a` and 1 optional argument `/r`.
+
+Example:
+```
+add income /a 100 /t salary /r 30
+```
+Below are the steps that shows the implementation of add income/expense.
+
+#### Step 1
+The AddCashflowCommand instance calls addIncome() or addExpense(), depending on what `category` is initialised as.
+
+addIncome() or addExpense() instantiates an Income or Expense object respectively.
+
+Example:
+```
+switch (category) {
+        case INCOME:
+            cashflowList.addIncome(amount, incomeType, recur);
+            break;
+        case EXPENSE:
+            cashflowList.addExpense(amount, expenseType, recur);
+            break;
+        default:
+            ui.showMessage("Unidentified entry.");
+            break;
+        }
+```
```

Comment:
Showing this using a diagram would be better, instead of showing this segment of your code.